### PR TITLE
migration: promote staged versions

### PR DESCRIPTION
**Pull request by bootjp**
## Summary
- add target-local PromotionState for staged MVCC promotion progress
- add memory and Pebble promotion primitives that copy staged rows to live keys and physically remove promoted staged rows in one apply
- add a Raft FSM command plus Internal.PromoteStagedVersions RPC
- preserve commit_ts, tombstones, and TTL metadata during promotion

## Tests
- go test ./store ./kv ./adapter -run 'Test(PromoteVersions|ApplyMigrationPromote|MigrationPromoteTargetKey|InternalPromoteStagedVersions|InternalImportRangeVersions|ExportVersions|ImportVersions|FSMApplyBatchKeepsPerRequestResults)'\n- go test ./store\n- go test ./kv\n- go test -run '^$' ./...\n- GOCACHE=$(pwd)/.cache GOLANGCI_LINT_CACHE=$(pwd)/.golangci-cache golangci-lint run ./store ./kv ./adapter --timeout=5m\n\nAuthor: bootjp

### CODE REVIEW

**Comment by bootjp** [comment]
@codex review

Validation for latest head 0df13a209a:
- go test ./store ./kv ./adapter -run 'Test(PromoteVersions|ApplyMigrationPromote|MigrationPromoteTargetKey|InternalPromoteStagedVersions|InternalImportRangeVersions|ExportVersions|ImportVersions|FSMApplyBatchKeepsPerRequestResults)'\n- go test ./store\n- go test ./kv\n- go test ./adapter -run 'TestInternal|TestDistribution|TestMigration'\n- go test -run '^$' ./...\n- golangci-lint run ./store ./kv ./adapter --timeout=5m

**Comment by bootjp** [comment]
@codex review

Validation for latest head 8c59f29f:
- go test ./store ./kv ./adapter -run 'Test(PromoteVersions|ApplyMigrationPromote|MigrationPromoteTargetKey|InternalPromoteStagedVersions|InternalImportRangeVersions|ExportVersions|ImportVersions|FSMApplyBatchKeepsPerRequestResults)'
- go test ./store
- go test ./kv
- go test -run '^$' ./...
- golangci-lint run ./store ./kv ./adapter --timeout=5m

File: store/migration_promote.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+	existing, ok := s.tree.Get(key)
+	if !ok {
+		return false
+	}
+	versions, _ := existing.([]VersionedValue)
+	idx := findVersionIndex(versions, commitTS)
+	if idx < 0 {
+		return false
+	}
+	versions = append(versions[:idx], versions[idx+1:]...)
```
[the remark this answers is not in the document]
Comment (bootjp) [reply store/migration_promote.go]: Fixed in b4f24775 by allocating a fresh version slice before removing the promoted staged version, so the tree-held backing array is not mutated in place.

**Comment by bootjp** [comment]
@codex review

Validation for latest head b4f24775:
- go test ./store -run 'TestPromoteVersions'
- golangci-lint run ./store ./kv ./adapter --timeout=5m

**Comment by bootjp** [comment]
Addressed latest promotion applied-index feedback in c1afe270:
- thread the Raft apply index from migration promotion apply into PromoteVersionsOptions
- persist metaAppliedIndex in the same Pebble batch as promoted live versions, staged deletes, and promotion state
- add Pebble coverage for promotion batches updating LastAppliedIndex

Validation:
- go test ./store -run 'Test(PromoteVersions|ApplyMutationsRaftAt|DeletePrefixAtRaftAt)'\n- go test ./kv -run 'TestApplyMigrationPromote'\n- go test ./store ./kv\n- go test -run '^$' ./...\n- GOCACHE=$(pwd)/.cache GOLANGCI_LINT_CACHE=$(pwd)/.golangci-cache golangci-lint run ./store ./kv --timeout=5m\n- git diff --check\n\n@codex review

**Comment by bootjp** [comment]
Addressed current-head promotion blockers in dd52db81988f75ecc1c7718935ac897dc2b52ac4.

Changes:
- completed Pebble promotion retries now persist metaAppliedIndex when AppliedIndex is provided
- PromoteStagedVersions is gated before proposing 0x0b entries; default gate stays closed until ELASTICKV_ENABLE_MIGRATION_PROMOTE_OPCODE is enabled after rollout
- migration promote apply failures now return HaltApply-marked errors instead of ordinary apply errors

Validation:
- go test ./store -run 'TestPromoteVersions_BundlesMetaAppliedIndex|Test(PromoteVersions|MigrationPromotion)'
- go test ./kv -run 'TestApplyMigrationPromote|TestMigrationPromote'
- go test ./adapter -run 'TestInternalPromoteStagedVersions|TestInternalImportRangeVersions'
- go test ./store -run '^TestPebbleStore_RestoreWaitsForMaintenanceLock$' -count=1
- go test ./adapter -run '^TestInternalPromoteStagedVersions' -count=1
- git diff --check

Note: attempted go test ./store ./kv ./adapter; kv passed, store hit a timeout in TestPebbleStore_RestoreWaitsForMaintenanceLock during the combined run, and that exact test passed when rerun alone.

@codex review

**Comment by bootjp** [comment]
Addressed current-head P1 promotion gate feedback in b04d30fe3cfe3dab6bd7bfdadf07920d1a886681.

Changes:
- threaded the writer-registration gate choice through applyImportVersionsBatch
- kept migration import on its existing gate behavior
- promotion now applies promoted target versions with gateRegistration=false so Raft replay is not blocked by this node's local writer-registration state
- added encrypted registration-gate regression coverage for PromoteVersions

Validation:
- go test ./store -run '^TestRegistrationGate_PromoteVersionsNeverGated$' -count=1
- go test ./store -run 'TestRegistrationGate|TestPromoteVersions|TestMigrationPromotion|TestPromoteVersions_BundlesMetaAppliedIndex' -count=1
- go test ./kv -run 'TestApplyMigrationPromote|TestMigrationPromote' -count=1
- go test ./adapter -run 'TestInternalPromoteStagedVersions|TestInternalImportRangeVersions' -count=1
- go test -run '^$' ./...
- golangci-lint run ./... --timeout=5m
- git diff --check

Note: attempted go test ./...; adapter package timed out after 10m while unrelated full-package network/raft adapter tests were still running. The targeted adapter migration tests above passed, and the all-package compile-only check passed.

@codex review

**Comment by bootjp** [comment]
Addressed current-head promote cursor feedback in c72eace4cd5007b515cda3d7fb961b4893e55e48.

Changes:
- validate PromoteStagedVersions cursor decode/range before proposing the Raft command
- reject malformed or wrong-job staged cursors as InvalidArgument
- keep proposer calls at zero for invalid cursor requests

Validation:
- go test ./adapter -run '^TestInternalPromoteStagedVersions' -count=1\n- go test ./store -run 'Test(PromoteVersions|MigrationPromotion)' -count=1\n- go test ./kv -run 'TestApplyMigrationPromote|TestMigrationPromote' -count=1\n- golangci-lint --config=.golangci.yaml run ./adapter ./store ./kv --timeout=5m\n- git diff --check\n\n@codex review

**Comment by bootjp** [comment]
Merged the cross-group migration base into the promote branch and resolved the pending merge state.

Validation:
- go test ./store ./kv ./adapter -run 'Test(PromoteVersions|ApplyMigrationPromote|MigrationPromoteTargetKey|InternalPromoteStagedVersions|InternalImportRangeVersions|ExportVersions|ImportVersions|FSMApplyBatchKeepsPerRequestResults|ShardStoreCommittedVersionAt|RejectsOnePhaseTxnBelowRouteFloor|Migration)' -count=1 -timeout=300s
- go test -run '^$' ./... -count=1 -timeout=300s
- golangci-lint --config=.golangci.yaml run ./store ./kv ./adapter --timeout=5m
- git diff --cached --check

@codex review

**Comment by bootjp** [comment]
Fixed the current-head P1 watermark blocker in 3cb3aad44078caf0a86233faf4dbac742fdc915d.

Validation:
- git diff --check
- go test ./store -run 'TestPebblePromoteVersionsAdvancesLastCommitTS|TestPromoteVersionsMovesStagedVersionsAndDeletesStagedRows|TestPromoteVersions_BundlesMetaAppliedIndex|TestRegistrationGate_PromoteVersionsNeverGated'
- go test ./store
- go test ./kv -run 'TestApplyMigrationPromoteMovesStagedVersions|TestMigrationPromoteTargetKeyRestoresRawKey'
- go test ./adapter -run 'TestInternalPromoteStagedVersions'
- GOCACHE=$(pwd)/.cache GOLANGCI_LINT_CACHE=$(pwd)/.golangci-cache golangci-lint run ./... --timeout=5m

Additional full sweep:
- go test ./... was attempted; github.com/bootjp/elastickv/adapter hit the default 10m test timeout. The promotion-specific adapter target above passed.

@codex review

**Comment by bootjp** [comment]
Fixed the promotion apply write-mode issue.

Summary:
- Threaded promotion write options through the Pebble promotion batch commit.
- `PromoteVersions` now uses `raftApplyWriteOpts()` when `AppliedIndex > 0` and keeps `directApplyWriteOpts()` for direct calls without a raft applied index.

Validation:
- `go test ./store -run 'TestPromoteVersionsWriteOptsFollowApplyContext|TestPromoteVersionsRaftNoSyncFunctionalEquivalence|TestPromoteVersions_BundlesMetaAppliedIndex|TestPromoteVersionsMovesStagedVersionsAndDeletesStagedRows|TestPebblePromoteVersionsAdvancesLastCommitTS|TestDirectApplyWriteOpts_AlwaysSync'`
- `go test ./store`
- `go test ./adapter -run 'TestInternalPromoteStagedVersions'`
- `golangci-lint run ./... --timeout=5m`
- `git diff --check`
- `go test ./...` was attempted; it timed out in `github.com/bootjp/elastickv/adapter` after 10m. The focused store tests and targeted migration-promote adapter tests passed.

Caller audit:
- Production promotion still enters through `adapter/internal.go -> kv.MarshalMigrationPromoteCommand -> kvFSM.applyMigrationPromote`, which passes `f.pendingApplyIdx` as `AppliedIndex`.
- Direct/test `PromoteVersions` callers leave `AppliedIndex=0`, so they keep synchronous direct commits.

@codex review

**Comment by bootjp** [comment]
Fixed the current ubuntu test failure in a9c9afef3b080c6979440b1b703faea673eebd9b.

Cause:
- `go test -race ./...` timed out in adapter while `TestLuaNegativeTypeCache_BoundedSize` was running for more than 9 minutes. The test was doing 10k+ real missing-key storage probes even though the invariant under test is the in-memory negative-cache bound.

Changes:
- kept `keyType` behavior unchanged by extracting the negative-cache insertion into `rememberNegativeType`.
- changed the bounded-size regression to exercise that in-memory bound directly, while leaving the wire-level Lua tests and single-probe test to cover real storage probing.

Validation:
- go test ./adapter -run 'TestLuaNegativeTypeCache_BoundedSize|TestLuaNegativeTypeCache_SingleProbePerKey|TestInternalPromoteStagedVersions' -count=1 -timeout=180s\n- go test -race ./adapter -run 'TestLuaNegativeTypeCache_BoundedSize|TestLuaNegativeTypeCache_SingleProbePerKey' -count=1 -timeout=180s\n- go test ./store -run 'TestPromoteVersionsWriteOptsFollowApplyContext|TestPromoteVersionsRaftNoSyncFunctionalEquivalence|TestPromoteVersions_BundlesMetaAppliedIndex|TestPromoteVersionsMovesStagedVersionsAndDeletesStagedRows|TestPebblePromoteVersionsAdvancesLastCommitTS|TestDirectApplyWriteOpts_AlwaysSync' -count=1 -timeout=180s\n- golangci-lint --config=.golangci.yaml run ./adapter ./store --timeout=5m\n- git diff --cached --check\n\n@codex review

**Comment by bootjp** [comment]
Resolved the merge conflict against design/hotspot-split-m2-cross-group in 5254a4330d9db25f4d9f8dd9ce2e77480de80442.

Integrated the newer exact-key migration metadata/export cursor changes with the staged promotion implementation by keeping promotion state under the reserved _migpromote metadata key and preserving the normalized export cursor path for promotion scans.

Validation:
- go test ./store -run 'Test(ExportVersions|PebbleExport|ImportVersions|PromoteVersions|PebblePromoteVersions|PebbleImportMetadataPersistsAcrossReopen|PebbleSnapshotPreservesMigrationMetadata|PebbleWriterRegistry|PebbleRejectsMVCCKey)' -count=1 -timeout=240s
- go test ./store ./kv ./adapter ./distribution ./internal/backup ./internal/s3keys -count=1 -timeout=300s (store/kv/distribution/internal passed; full adapter package timed out)
- go test ./adapter -run 'Test(Milestone1SplitRange|DistributionServer|BuildCatalogSplitOps|GRPCServer_Raw|Internal(ExportRangeVersions|ImportRangeVersions|PromoteStagedVersions))' -count=1 -timeout=240s
- golangci-lint --config=.golangci.yaml run ./store ./kv ./adapter ./distribution ./internal/backup ./internal/s3keys --timeout=5m
- git diff --check HEAD

@codex review

**Comment by bootjp** [comment]
Addressed the current-head promotion blockers in b1381f3afb46d4a7c05f56427b539b94c35c0fa2.

Changes:
- reject skipped-key cursors inside the requested range and restrict promotion resume cursors to empty or emitted positions before proposing to Raft
- reject scanned/pruned/skipped promotion cursors at the Internal RPC boundary with InvalidArgument
- clear restored in-memory promotion state during MVCC snapshot restore and cover same-job resume after restore

Validation:
- go test ./store ./adapter -run 'Test(ValidateExportCursorForRangeRejectsSkippedCursorInsideRange|ValidatePromotionCursorForRangeAcceptsOnlyEmittedPositions|MVCCStore_RestoreClearsMigrationMetadata|InternalPromoteStagedVersionsRejectsInvalidCursorBeforePropose)' -count=1 -timeout=240s
- go test ./store ./kv ./adapter -run 'Test(ExportVersions|ValidateExportCursor|ValidatePromotionCursor|PromoteVersions|PebblePromoteVersions|MVCCStore_RestoreClearsMigrationMetadata|ApplyMigrationPromote|MigrationPromote|InternalPromoteStagedVersions)' -count=1 -timeout=240s
- go test ./store -count=1 -timeout=300s
- golangci-lint --config=.golangci.yaml run ./store ./adapter --timeout=5m
- git diff --check

@codex review

**Comment by bootjp** [comment]
@codex review

Merged latest design/hotspot-split-m2-cross-group into this branch in b4b8664c84d8c3c054d5ab33b7631023fea0ec53 and resolved the adapter/internal.go + adapter/internal_migration_test.go conflicts.

Conflict resolution:
- kept promote opcode gate and promotion cursor validation from this branch.
- kept import job_id/bracket_id validation and S3 auxiliary route/export fixes from the updated base.
- retained staged S3 auxiliary routing/read-key test coverage and promotion restore metadata tests.

Validation:
- go test ./adapter ./kv ./store -run "Test(Internal(ExportRangeVersions|ImportRangeVersions|PromoteStagedVersions)|Validate(Export|Promotion)Cursor|PromoteVersions|PebblePromoteVersions|ApplyMigration(Promote|Import)|Migration(Promote|Import)|ShardedCoordinatorRoutesS3BucketAuxiliaryWriteToStagedOwner|GroupReadKeysByShardID_RoutesS3BucketAuxiliaryToStagedOwner|ShardStore.*Staged|MVCCStore_RestoreClearsMigrationMetadata)" -count=1 -timeout=240s
- golangci-lint --config=.golangci.yaml run ./adapter ./kv ./store --timeout=5m
- git diff --check / git diff --cached --check

**Comment by bootjp** [comment]
@codex review

Addressed current-head stale promotion cursor handling in 40c2f527b3fd6d10d22e633112d703de373c6f6a.

Changes:
- job-scoped memory/Pebble promotion now restart from the target-local state when present and from an empty cursor when state is missing, instead of trusting a stale caller cursor.
- added memory/Pebble regression coverage that passes a stale in-range cursor with no PromotionState and verifies all staged rows are promoted.

Validation:
- go test ./store -run 'Test(PromoteVersionsIgnoresClientCursorWhenStateMissing|PromoteVersionsMovesStagedVersionsAndDeletesStagedRows|PebblePromoteVersionsAdvancesLastCommitTS|ValidatePromotionCursorForRangeAcceptsOnlyEmittedPositions)' -count=1 -timeout=240s
- go test ./store -count=1 -timeout=300s
- go test ./adapter ./kv ./store -run 'Test(InternalPromoteStagedVersionsRejectsWhenOpcodeGateClosed|InternalPromoteStagedVersionsRejectsInvalidCursorBeforePropose|ApplyMigrationPromote|MigrationPromote|ValidateExportCursorForRangeRejectsSkippedCursorInsideRange|ValidatePromotionCursorForRangeAcceptsOnlyEmittedPositions|PromoteVersionsIgnoresClientCursorWhenStateMissing|PromoteVersionsMovesStagedVersionsAndDeletesStagedRows|PebblePromoteVersionsAdvancesLastCommitTS|MVCCStore_RestoreClearsMigrationMetadata)' -count=1 -timeout=240s
- golangci-lint --config=.golangci.yaml run ./store --timeout=5m
- git diff --check / git diff --cached --check

**Comment by bootjp** [comment]
@codex review

Synced PR #1089 with latest origin/design/hotspot-split-m2-cross-group at 19a28e908e5ebcff0b8c04172f5bd0a3297a4ff5 via merge commit 327cb1c8f72fbe07e893bf24fd24c1b2ed2d7ea5.

Current head includes the stale promotion cursor fix from 40c2f527b3fd6d10d22e633112d703de373c6f6a:
- job-scoped memory/Pebble promotion now uses target-local PromotionState when present and restarts from an empty cursor when that state is missing, instead of trusting a stale caller cursor.
- regression coverage verifies stale in-range client cursors do not skip staged rows when PromotionState is absent.

Validation after the base merge:
- go test ./adapter ./kv ./store -run 'Test(InternalPromoteStagedVersionsRejectsWhenOpcodeGateClosed|InternalPromoteStagedVersionsRejectsInvalidCursorBeforePropose|ApplyMigrationPromote|MigrationPromote|ValidateExportCursorForRangeRejectsSkippedCursorInsideRange|ValidatePromotionCursorForRangeAcceptsOnlyEmittedPositions|PromoteVersionsIgnoresClientCursorWhenStateMissing|PromoteVersionsMovesStagedVersionsAndDeletesStagedRows|PebblePromoteVersionsAdvancesLastCommitTS|MVCCStore_RestoreClearsMigrationMetadata|ShardedCoordinatorRoutesS3BucketAuxiliaryWriteToStagedOwner|GroupReadKeysByShardID_RoutesS3BucketAuxiliaryToStagedOwner|ShardStore(GetAt_MergesStagedVisibilityForS3BucketAuxiliary|ScanAt_ContinuesStagedVisibilityAfterCandidateWindow|ScanAtRestrictsStagedVisibilityToSafeFrontier|ExplicitGroupReads_MergeStagedVisibility))' -count=1 -timeout=240s
- go test ./kv -count=1 -timeout=300s
- golangci-lint --config=.golangci.yaml run ./kv ./store --timeout=5m
- git diff --check

**Comment by bootjp** [comment]
@codex review

Merged latest design/hotspot-split-m2-cross-group into this branch in 4b1d950f975de31551ce8789b950aff92ed5e260 after the #1088 base advanced.

Validation:
- go test ./store ./kv -run "Test(PromoteVersionsIgnoresClientCursorWhenStateMissing|PromoteVersionsMovesStagedVersionsAndDeletesStagedRows|PebblePromoteVersionsAdvancesLastCommitTS|ValidatePromotionCursorForRangeAcceptsOnlyEmittedPositions|ShardStore.*Raft|ApplyMutationsRaft|DeletePrefixAtRaft)" -count=1 -timeout=240s
- go test ./adapter ./distribution -run "Test(DistributionServer(GetRouteOwnership|GetIntersectingRoutes|OwnershipRPCs|SplitRange)|RouteHistorySnapshot|EngineSnapshotAt|EngineGetIntersectingRoutes)" -count=1 -timeout=240s
- golangci-lint --config=.golangci.yaml run ./adapter ./distribution ./kv ./store --timeout=5m
- git diff --check

**Comment by bootjp** [comment]
@codex review

Addressed current-head promotion/export review feedback in f902f34dab2211d0b5e2731d2c5d326e00fba8aa.

Changes:
- persist per-job promotion MaxPromotedTS in PromotionState and return it on completed-job retries
- use the persisted promotion timestamp floor when Pebble refreshes metaLastCommitTS on retry
- normalize skipped export cursors below StartKey so range exports resume at StartKey instead of scanning from the skipped key
- add regression coverage for skipped-cursor range resume, promotion retry floor preservation, Pebble reopen retry behavior, and promotion-state codec compatibility

Validation:
- go test ./store -run 'Test(ExportVersionsSkippedCursorBeforeStartResumesAtStartKey|ValidatePromotionCursorForRangeAcceptsOnlyEmittedPositions|PromoteVersionsMovesStagedVersionsAndDeletesStagedRows|PromoteVersionsIgnoresClientCursorWhenStateMissing|PebblePromoteVersionsAdvancesLastCommitTS|PromotionStateCodecPreservesMaxPromotedTS)$' -count=1 -timeout=240s
- go test ./kv -run 'TestApplyMigrationPromote|TestMigrationPromote' -count=1 -timeout=240s
- go test ./adapter -run 'TestInternalPromoteStagedVersions' -count=1 -timeout=240s
- go test ./store -run 'Test(PromoteVersions|PebblePromoteVersions|MigrationPromotion|PebbleStore_RestoreWaitsForMaintenanceLock|Pebble.*AppliedIndex|Pebble.*Registration|Pebble.*Sync)' -count=1 -timeout=300s
- go test ./store ./kv -count=1 -timeout=300s
- go test ./adapter -run 'TestInternalPromoteStagedVersions|TestInternalImportRangeVersions|TestInternalExportVersions' -count=1 -timeout=240s
- golangci-lint --config=.golangci.yaml run ./store ./kv ./adapter --timeout=5m
- git diff --check

Note: full go test ./store ./kv ./adapter reached the 300s package timeout in adapter after store and kv had completed successfully, so I used the successful focused adapter migration coverage above for this change.

**Comment by bootjp** [comment]
@codex review

Addressed the latest promote-cursor apply feedback in e3e37cebc9328ff9dc2102ec9dd3648b93fd038e.

Changes:
- keep committed migration-promote ErrInvalidExportCursor results as ordinary apply errors instead of HaltApply responses
- preserve HaltApply behavior for malformed promote payloads, unsupported stores, and non-validation storage failures
- added FSM coverage that invalid promote cursors do not flow through HaltApply
- extended promotion FSM coverage to assert the durable promoted timestamp floor

Validation:
- go test ./kv -run 'TestApplyMigrationPromote' -count=1 -timeout=240s
- go test ./kv -run 'TestApplyMigrationPromote|TestMigrationPromote' -count=1 -timeout=240s
- go test ./store -run 'Test(ExportVersionsSkippedCursorBeforeStartResumesAtStartKey|ValidatePromotionCursorForRangeAcceptsOnlyEmittedPositions|PromoteVersionsMovesStagedVersionsAndDeletesStagedRows|PromoteVersionsIgnoresClientCursorWhenStateMissing|PebblePromoteVersionsAdvancesLastCommitTS|PromotionStateCodecPreservesMaxPromotedTS|PromoteVersions|PebblePromoteVersions|MigrationPromotion|PebbleStore_RestoreWaitsForMaintenanceLock|Pebble.*AppliedIndex|Pebble.*Registration|Pebble.*Sync)$' -count=1 -timeout=300s
- go test ./adapter -run 'TestInternalPromoteStagedVersions|TestInternalImportRangeVersions|TestInternalExportVersions' -count=1 -timeout=240s
- go test ./store ./kv -count=1 -timeout=300s
- golangci-lint --config=.golangci.yaml run ./store ./kv ./adapter --timeout=5m
- git diff --check

File: store/migration_versions.go [hunk 2]
```diff
@@ -89,6 +91,25 @@ func decodeExportCursor(cursor []byte) (exportCursorPosition, error) {
...
+	if !pos.hasKey {
+		return nil
+	}
+	if startKey != nil && bytes.Compare(pos.key, startKey) < 0 {
+		return errors.WithStack(ErrInvalidExportCursor)
+	}
+	if endKey != nil && bytes.Compare(pos.key, endKey) >= 0 {
+		return errors.WithStack(ErrInvalidExportCursor)
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (bootjp) [reply store/migration_versions.go:139]: Addressed in f902f34d: ValidatePromotionCursorForRange now accepts only emitted cursors before proposal, so skipped/scanned/pruned sparse-scan cursor tags are rejected as ErrInvalidExportCursor. Covered by TestValidatePromotionCursorForRangeAcceptsOnlyEmittedPositions and adapter pre-propose cursor validation.

File: store/migration_versions.go [hunk 3]
```diff
@@ -89,6 +91,54 @@ func decodeExportCursor(cursor []byte) (exportCursorPosition, error) {
...
+func validateExportCursorPositionForRange(pos exportCursorPosition, startKey, endKey []byte) error {
+	if !pos.hasKey {
+		return nil
+	}
+	if pos.tag == exportCursorTagSkippedKey {
+		opts := ExportVersionsOptions{StartKey: startKey, EndKey: endKey}
+		if !exportSkippedCursorOutsideRange(opts, pos.key) {
+			return errors.WithStack(ErrInvalidExportCursor)
+		}
+		return nil
```
[the remark this answers is not in the document]
Comment (bootjp) [reply store/migration_versions.go:131]: Addressed in f902f34d: decodeExportCursorForOptions normalizes a skipped cursor below StartKey to an empty cursor, so both memory and Pebble export paths seek from StartKey instead of resuming at the skipped key. Covered by TestExportVersionsSkippedCursorBeforeStartResumesAtStartKey.

File: store/migration_promote.go [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+	promoted.NextCursor = exported.NextCursor
+	promoted.Done = exported.Done
+	promoted.ScannedBytes = exported.ScannedBytes
+	promoted.TotalPromotedRows = promoted.PromotedRows
+	if opts.JobID == 0 {
+		return promoted, nil
+	}
+	state.Cursor = bytes.Clone(exported.NextCursor)
+	state.Done = exported.Done
+	state.PromotedRows += promoted.PromotedRows
```
[the remark this answers is not in the document]
Comment (bootjp) [reply store/migration_promote.go:311]: Addressed in f902f34d: PromotionState now persists MaxPromotedTS, PromoteVersions returns the durable floor on completed-job retries, and Pebble retries use it when refreshing metaLastCommitTS. Covered by TestPromoteVersionsMovesStagedVersionsAndDeletesStagedRows, TestPebblePromoteVersionsAdvancesLastCommitTS, and TestPromotionStateCodecPreservesMaxPromotedTS.
